File: lib/application/anki_official/render/official_anki_http_range.dart
/// Single HTTP Range parser. Multi-range is rejected with 416.
class OfficialAnkiHttpRange {
  const OfficialAnkiHttpRange._({
    required this.satisfiable,
    required this.start,
    required this.end,
    required this.length,
    required this.status,
  });

  factory OfficialAnkiHttpRange.unsatisfiable(int length) => OfficialAnkiHttpRange._(
        satisfiable: false,
        start: 0,
        end: -1,
        length: length,
        status: 416,
      );

  factory OfficialAnkiHttpRange.full(int length) => OfficialAnkiHttpRange._(
        satisfiable: true,
        start: 0,
        end: length <= 0 ? -1 : length - 1,
        length: length,
        status: 200,
      );

  final bool satisfiable;
  final int start;
  final int end;
  final int length;
  final int status;

  int get contentLength =>
      satisfiable && end >= start ? end - start + 1 : 0;

  String get contentRange =>
      satisfiable ? 'bytes $start-$end/$length' : 'bytes */$length';

  static OfficialAnkiHttpRange parse(String? header, int length) {
    if (header == null || header.trim().isEmpty) {
      return OfficialAnkiHttpRange.full(length);
    }
    final trimmed = header.trim();
    if (!trimmed.toLowerCase().startsWith('bytes=')) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    final spec = trimmed.substring(6).trim();
    if (spec.isEmpty || spec.contains(',')) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    final dash = spec.indexOf('-');
    if (dash < 0) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    final left = spec.substring(0, dash).trim();
    final right = spec.substring(dash + 1).trim();
    if (length <= 0) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    if (left.isEmpty && right.isEmpty) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    if (left.isEmpty) {
      final suffix = int.tryParse(right);
      if (suffix == null || suffix <= 0) {
        return OfficialAnkiHttpRange.unsatisfiable(length);
      }
      final start = length > suffix ? length - suffix : 0;
      return OfficialAnkiHttpRange._(
        satisfiable: true,
        start: start,
        end: length - 1,
        length: length,
        status: 206,
      );
    }
    final start = int.tryParse(left);
    if (start == null || start < 0 || start >= length) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    final end = right.isEmpty
        ? length - 1
        : int.tryParse(right);
    if (end == null || end < start) {
      return OfficialAnkiHttpRange.unsatisfiable(length);
    }
    final clampedEnd = end >= length ? length - 1 : end;
    return OfficialAnkiHttpRange._(
      satisfiable: true,
      start: start,
      end: clampedEnd,
      length: length,
      status: 206,
    );
  }

  List<int> slice(List<int> bytes) {
    if (!satisfiable || bytes.length != length) {
      return const <int>[];
    }
    if (end < start) return const <int>[];
    return bytes.sublist(start, end + 1);
  }
}

File: android/app/src/main/kotlin/me/dsdogs/turna/anki/reviewer/OfficialAnkiHttpRange.kt
package me.dsdogs.turna.anki.reviewer

data class OfficialAnkiHttpRange(
    val satisfiable: Boolean,
    val start: Long,
    val end: Long,
    val length: Long,
    val status: Int,
) {
    val contentLength: Long
        get() = if (satisfiable && end >= start) end - start + 1 else 0

    val contentRange: String
        get() = if (satisfiable) "bytes $start-$end/$length" else "bytes */$length"

    companion object {
        fun unsatisfiable(length: Long) = OfficialAnkiHttpRange(false, 0, -1, length, 416)

        fun full(length: Long) = OfficialAnkiHttpRange(
            true,
            0,
            if (length <= 0) -1 else length - 1,
            length,
            200,
        )

        fun parse(header: String?, length: Long): OfficialAnkiHttpRange {
            if (header.isNullOrBlank()) return full(length)
            val trimmed = header.trim()
            if (!trimmed.startsWith("bytes=", ignoreCase = true)) {
                return unsatisfiable(length)
            }
            val spec = trimmed.substring(6).trim()
            if (spec.isEmpty() || spec.contains(',')) return unsatisfiable(length)
            val dash = spec.indexOf('-')
            if (dash < 0) return unsatisfiable(length)
            val left = spec.substring(0, dash).trim()
            val right = spec.substring(dash + 1).trim()
            if (length <= 0) return unsatisfiable(length)
            if (left.isEmpty() && right.isEmpty()) return unsatisfiable(length)
            if (left.isEmpty()) {
                val suffix = right.toLongOrNull() ?: return unsatisfiable(length)
                if (suffix <= 0) return unsatisfiable(length)
                val start = if (length > suffix) length - suffix else 0
                return OfficialAnkiHttpRange(true, start, length - 1, length, 206)
            }
            val start = left.toLongOrNull() ?: return unsatisfiable(length)
            if (start < 0 || start >= length) return unsatisfiable(length)
            val end = if (right.isEmpty()) {
                length - 1
            } else {
                right.toLongOrNull() ?: return unsatisfiable(length)
            }
            if (end < start) return unsatisfiable(length)
            val clamped = if (end >= length) length - 1 else end
            return OfficialAnkiHttpRange(true, start, clamped, length, 206)
        }
    }
}
